Kernel/Time/TimeManagement.cpp

```diff
@@ -0,0 +1,251 @@
+            if (m_hardware_timers[index]->is_periodic_capable()) {
+                periodic_timers_indexes.append(index);
+                if (enable_periodic_mode)
+                    m_hardware_timers[index]->set_periodic();
+            }
+        }
+    }
+    return periodic_timers_indexes;
+}
+
+Vector<size_t> TimeManagement::scan_for_non_periodic_timers()
+{
+    dbg() << "Scanning for Non-Periodic timers";
+    Vector<size_t> non_periodic_timers_indexes;
+    non_periodic_timers_indexes.ensure_capacity(m_hardware_timers.size());
+    for (size_t index = 0; index < m_hardware_timers.size(); index++) {
+        if (!m_hardware_timers[index].is_null())
+            if (!m_hardware_timers[index]->is_periodic_capable())
+                non_periodic_timers_indexes.append(index);
+    }
+    return non_periodic_timers_indexes;
+}
+
+bool TimeManagement::is_hpet_periodic_mode_allowed()
+{
+    if (!KParams::the().has("hpet")) {
+        return false;
+    }
+    auto hpet_mode = KParams::the().get("hpet");
+    if (hpet_mode == "periodic")
+        return true;
+    if (hpet_mode == "nonperiodic")
+        return false;
+    ASSERT_NOT_REACHED();
+}
+
+bool TimeManagement::probe_and_set_non_legacy_hardware_timers()
+{
+    if (!ACPI::Parser::the().is_operable())
+        return false;
+    if (!HPET::test_and_initialize())
+        return false;
+    if (!HPET::the().comparators().size()) {
+        dbg() << "HPET initialization aborted.";
+        return false;
+    }
+    dbg() << "HPET: Setting appropriate functions to timers.";
+
+    m_hardware_timers.resize(HPET::the().comparators().size());
+    for (size_t index = 0; index < m_hardware_timers.size(); index++) {
+        m_hardware_timers[index] = HPET::the().comparators()[index];
+#ifdef TIME_DEBUG
+        dbg() << m_hardware_timers[index].ptr() << " <- " << HPET::the().comparators()[index].ptr();
+#endif
+    }
+
+    auto periodic_timer_indexes = scan_and_set_periodic_timers();
+    auto non_periodic_timer_indexes = scan_for_non_periodic_timers();
+
+    if (is_hpet_periodic_mode_allowed())
+        ASSERT(!periodic_timer_indexes.is_empty());
+
+    if (!non_periodic_timer_indexes.is_empty())
+        m_system_timer = m_hardware_timers[non_periodic_timer_indexes[0]];
+    else
+        m_system_timer = m_hardware_timers[periodic_timer_indexes[1]];
+
+    m_time_keeper_timer = m_hardware_timers[periodic_timer_indexes[0]];
+
+    m_system_timer->change_function([&](const RegisterState& regs) { update_scheduler_ticks(regs); });
+    m_system_timer->try_to_set_frequency(OPTIMAL_TICKS_PER_SECOND_RATE);
+    m_time_keeper_timer->change_function([&](const RegisterState& regs) { update_time(regs); });
+    m_time_keeper_timer->try_to_set_frequency(OPTIMAL_TICKS_PER_SECOND_RATE);
+
+    return true;
+}
+
+bool TimeManagement::probe_and_set_legacy_hardware_timers()
+{
+    if (ACPI::Parser::the().is_operable()) {
+        if (ACPI::Parser::the().x86_specific_flags().cmos_rtc_not_present) {
```

Comment:
Due to a bug in QEMU, even when ACPI FADT version is above 2 (which means that IA-PC flags are valid), I verified that and the emulator simply puts a zero, regardless if CMOS RTC is supported or not. This bug can be classified as a firmware bug on real machines, and we should tolerate it somehow.

More details here:
https://forum.osdev.org/viewtopic.php?f=1&t=30861
https://github.com/qemu/qemu/blob/master/hw/acpi/aml-build.c#L1820
https://github.com/qemu/qemu/commit/8612f8bd9fb667c62e9a56c67ab2e3af2798dd9c

Reply:
It's possible that we could tolerate misleading IA-PC flags' bitfield with the ACPI namespace, so if we enumerate the AML bytecode (we don't that yet), we can find what is present on the ISA bus on real hardware or emulated hardware. Problem is that if the user specified the boot argument `acpi=limited`, we can't enumerate AML, thus, we can't avoid a firmware bug like this one.